### ISSUE TITLE
nokogiri java does not properly handle xml files with byte order markers

### ISSUE THREAD

**Issue by nc-crowec** [sparklemotion/nokogiri#1768]
We're updating our app from CRuby to JRuby, so we're swapping to the Java version of the nokogiri gem. I'm seeing some test failures, and the root cause seems to be that nokogiri was happy to work with XML files/schemas contaiing byte-order-markers before, but the java version does not like them. An error is occurring in the java extensions. The issue can be reproduced with the following script - which works fine on the non-java version of nokogiri, but produces errors or otherwise misbehaves on the java version. Tested with 1.8.3.

```
# create a string with only the UTF-8 BOM
bom = [239, 187, 191].pack('c*').force_encoding('UTF-8')

schema = bom + %Q(<?xml version = "1.0" encoding = "UTF-8"?>
<xs:schema xmlns:xs = "http://www.w3.org/2001/XMLSchema">
   <xs:element name = "contact">
      <xs:complexType>
         <xs:sequence>
            <xs:element name = "name" type = "xs:string" />
         </xs:sequence>
      </xs:complexType>
   </xs:element>
</xs:schema>
)

Nokogiri::XML.Schema(schema)
```

Which results in an error like the following:
```
Nokogiri::XML::SyntaxError: Content is not allowed in prolog.
	from nokogiri/XmlSchema.java:134:in `from_document'
```

When trying to parse XML with a BOM, it does not barf, but the document appears empty:

```
# create a string with only the UTF-8 BOM
bom = [239, 187, 191].pack('c*').force_encoding('UTF-8')

xml = bom + %Q(<?xml version = "1.0" encoding = "UTF-8"?>
<book>
  <title>Learning XML</title>
  <price>39.95</price>
</book>
)

Nokogiri::XML(xml)
```

Which results in:
```
 => #<Nokogiri::XML::Document:0x7d2 name="document"> 
```

I understand that we can strip the BOMs to work around this issue - but as the CRuby version of the gem seems to handle them just fine, I'm curious if the JRuby version can be likewise updated?

**Comment by flavorjones** [comment]
Thanks for reporting this. I'm tagging some of the people who know the Java implementation better than I do, @jvshahid and @kares, to see if they have some time to look into it.

**Comment by jvshahid** [comment]
@corycrowe how are you getting strings with BOM in them ? are those strings that you read from disk ? If so, can you use `File.read(path, "bom|utf-8")` to let JRuby/MRI strip the BOM before parsing the content. 

For context, Java doesn't handle BOM as part of encoding (there are a couple of issues that were marked as `wontfix` with backward compatibility cited as the reason). I'm not sure which part on MRI is stripping the BOM, but libxml2 support BOM stripping as part of the internal encoding code. I think we could use something like [BOMInputStream](http://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/input/BOMInputStream.html) but I would like to justify the change given that Ruby have tools to strip those BOM for us.

**Comment by jvshahid** [comment]
we try to keep the two implementation consistent as much as we can. obviously this isn't always possible, since doing so would require non trivial changes sometimes. That said, I think this situation can be fixed in Java by wrapping the input streams that gets passed to the parser [here](https://github.com/sparklemotion/nokogiri/blob/117ca2e067dbbf054bef9078c79387c8170d2156/ext/java/nokogiri/internals/ParserContext.java#L88). I think if we decide to go down this path, we will have to be careful to not throw the BOM away and instead try to figure out the encoding and pass that through to the parser.

**Comment by kares** [comment]
@jvshahid @flavorjones do you guys happen to know if BOM removal happens at C-Ruby level or libxml?
... there's some support for BOM skips in JRuby so it might be a question of using more of the native APIs

**Comment by jvshahid** [comment]
I haven't looked deeply at the C-Ruby implementation but I saw in the libxml2 docs that BOM is supported and used to set the encoding of the input stream. I can look into this later.

**Comment by flavorjones** [comment]
Heyo, libxml2 natively handles BOM via the method `xmlDetectCharEncoding` which is invoked by both the xml and html parsers. Nokogiri's C-Ruby implementation doesn't know anything about BOM (and neither do I, really!).

**Comment by flavorjones** [comment]
This was fixed by ba16682aae2cbd42c196bd8afdfcfe8a5d82fbdb in v1.11.0 (2021-01-03).
